runtime/vm/zcinterp.m4

```diff
@@ -153,70 +153,101 @@ END_CURRENT
 })
 
 ifdef({OMR_GC_CONCURRENT_SCAVENGER},{
-dnl Helper to handle guarded storage events and software read barriers. When
-dnl the event occurs, hardware/software populates the return address in 
-dnl J9VMThread.gsParameters.returnAddr.
-define({HANDLE_GS_EVENT_HELPER},{
-    SAVE_ALL_REGS($1)
-    ST_GPR J9SP,J9TR_VMThread_sp(J9VMTHREAD)
-    LR_GPR CARG1,J9VMTHREAD
-    L_GPR CRA,J9TR_VMThread_javaVM(J9VMTHREAD)
-    L_GPR CRA,J9TR_JavaVM_invokeJ9ReadBarrier(CRA)
-    CALL_INDIRECT(CRA)
-    L_GPR J9SP,J9TR_VMThread_sp(J9VMTHREAD)
-    ST_GPR J9SP,JIT_GPR_SAVE_SLOT(J9SP)
-    RESTORE_ALL_REGS_AND_SWITCH_TO_JAVA_STACK($1)
-})
+    define({HANDLE_HARDWARE_READ_BARRIER},{
+    BEGIN_HELPER($1)
+        SAVE_ALL_REGS($1)
 
-define({HANDLE_GS_EVENT},{
-BEGIN_FUNC($1)
 dnl Test the TX bit of GSECI (bit 16) to see if the CPU was in transactional-
 dnl execution mode when the guarded storage event was recognized. At that point
 dnl we have two cases:
 dnl
 dnl 1. TX was a non-constrained transaction:
-dnl	   If this was the case do not execute the read barrier and simply return to
+dnl    If this was the case do not execute the read barrier and simply return to
 dnl    the fallback transaction abort JIT code to handle it by forcing condition
 dnl    code to be 2 with the below CL(G)FI instruction indicating a transitent 
 dnl    transaction abort.
 dnl
 dnl 2. TX was a constrained transaction:
-dnl    If this was the case then we are safe to execute the read barrier because
-dnl    the JIT has ensured that there will be no side-effects of the transaction
-dnl    being aborted prior to reaching the guarded storage event handler. In this
-dnl    case it is functionally safe to execute the read barrier.
+dnl    This should not happen because we cannot handle constrained transactions
+dnl    in the hardware read barrier case. Because executing any of the following
+dnl    code will be functionally incorrect we force a crash here so we fail fast
+dnl    and can manually debug.
 dnl
-dnl See eclipse/openj9#2545 and eclipse/openj9#2790 for more details.
-    TM J9TR_VMThread_gsParameters_GSECI(J9VMTHREAD),128
-    JZ LABEL_NAME(L_GS_CALLHELPER)
-    TM J9TR_VMThread_gsParameters_GSECI(J9VMTHREAD),64
-    JNZ LABEL_NAME(L_GS_CALLHELPER)
-    CLFI_GPR J9SP,0
-    J LABEL_NAME(L_GS_DONE)
-PLACE_LABEL(L_GS_CALLHELPER)
-    RELOAD_SSP_SAVE_VALUE
-    STM_GPR r0,LAST_REG,JIT_GPR_SAVE_SLOT(0)
-    SWAP_SSP_VALUE
-    SAVE_HIWORD_REGS
-    LOAD_CAA
-    HANDLE_GS_EVENT_HELPER($1)
-PLACE_LABEL(L_GS_DONE)
-dnl Branch back to the instruction that triggered guarded storage event.
-    BRANCH_INDIRECT_ON_CONDITION(15,J9TR_VMThread_gsParameters_returnAddr,0,J9VMTHREAD)
-END_CURRENT
-})
+dnl See the following PRs for more details:
+dnl eclipse/openj9#2545
+dnl eclipse/openj9#2790
+dnl eclipse/openj9#9999
```

Comment:
Uuups. Just realized I left this as an eyecatcher because I didn't know what the PR ID was going to be before opening this. Fixing this momentarily.

Reply:
Fixed in 6260445